universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,206 @@
+    //if (auto* drefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<double>(name)) // TODO C++17
+    if (drefp)
+        return drefp;
+    auto* irefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<int>(name);
+    //if (auto* irefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<int>(name)) // TODO C++17
+    if (irefp)
+        return irefp;
+    const auto& it = m_value_refs.find(name);
+    return it != m_value_refs.end() ? it->second.get() : nullptr;
+}
+
+NamedValueRefManager& NamedValueRefManager::GetNamedValueRefManager() {
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager starts (check the thread)"; 
+    
+    static NamedValueRefManager manager; // function local 
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager at " << &manager;
+    return manager;
+}
+
+
+unsigned int NamedValueRefManager::GetCheckSum() const {
+    unsigned int retval{0};
+    for (auto const& name_type_pair : m_value_refs)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    for (auto const& name_type_pair : m_value_refs_int)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    for (auto const& name_type_pair : m_value_refs_double)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    //CheckSums::CheckSumCombine(retval, m_value_refs.size()); // why also add size?
+    
+    DebugLogger() << "NamedValueRefManager checksum: " << retval;
+    return retval;
+}
+
+
+NamedValueRefManager::any_container_type  NamedValueRefManager::GetItems() const {
+    std::function<const any_entry_type(const entry_type&)> base_to_base          = [](const entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+    std::function<const any_entry_type(const double_entry_type&)> double_to_base = [](const double_entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+    std::function<const any_entry_type(const int_entry_type&)> int_to_base       = [](const int_entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+
+    // should use C++20 ranges (or C++14, C++17 boost Range-v3 library) to avoid copying
+    any_container_type aet;
+    std::transform(m_value_refs_double.begin(), m_value_refs_double.end(), std::inserter(aet, aet.end()), double_to_base);
+    std::transform(m_value_refs_int.begin(), m_value_refs_int.end(), std::inserter(aet, aet.end()), int_to_base);
+    std::transform(m_value_refs.begin(), m_value_refs.end(), std::inserter(aet, aet.end()), base_to_base);
+    for (auto kv : aet)
+        ErrorLogger() << "NamedValueRefManager::GetItems - " << kv.first << " " << kv.second.get().StringResult() << " := " << kv.second.get().Description() ;
+    return aet;
+}
+
+template <typename R, typename VR>
+void NamedValueRefManager::RegisterValueRefImpl(R& container, std::mutex& mutex, const std::string&& label, std::string&& valueref_name, std::unique_ptr<VR>&& vref) {
```

Comment:
`const std::string&& label` ?

Reply:
`const std::string& label` ?

Reply:
probably `const std::string& label`, `std::string label`, or `std::string&& label` depending how you use it...

Reply:
writing to log

Reply:
and what is passed... looks like a string literal so `const std::string&` is probably ok